go/vt/vtgate/safe_session.go

```diff
@@ -16,7 +16,7 @@ import (
 // (the use pattern is 'Find', if not found, then 'Append',
 // for a single shard)
 type SafeSession struct {
-	mu sync.Mutex
+	mu sync.RWMutex
```

Comment:
This helps with scatter queries (sharded goroutines access the same session). I noticed that sometimes sharded queries of a scatter query are not sent out at the same time. Though I did not investigate the cause so I cannot say this definitely contributes.

I agree it is not a global lock affecting multiple keyspaces. So I can revert this if you think it does not worth.